config-related test errors

When accessing configuration data using `spack.config.[gs]et` at some point before test setup, something goes wrong and the test configuration will not be set up properly, leading to test failures. I encountered this in #17229 because it introduces `spack.config.get` calls in Spack's main function. However, it is enough to enable debugging to reproduce the problem.

### Steps to reproduce the issue

```console
$ spack -d test -k test_exclude_specs
```

### Error Message

For some reason, the tests try to use my normal environment instead of the test environment:

```
E           spack.concretize.NoCompilersForArchError: No compilers found for operating system debian6 and target x86_64.
E           If previous installations have succeeded, the operating system may have been updated.
E           Compilers are defined for the following operating systems and targets:
E           	fedora32-x86_64
E               Run 'spack compiler find' to add compilers.
```

### Information on your system

* **Spack:** 0.14.2-1557-096bd69a9
* **Python:** 3.8.3
* **Platform:** linux-fedora32-haswell

### Additional information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have searched the issues of this repo and believe this is not a duplicate
- [x] I have run the failing commands in debug mode and reported the output

--- a/lib/spack/spack/test/conftest.py
+++ b/lib/spack/spack/test/conftest.py
@@ -268,6 +268,10 @@ def _skip_if_missing_executables(request):
 # FIXME: there's some weird interaction with compilers during concretization.
 spack.architecture.real_platform = spack.architecture.platform
 spack.architecture.platform = lambda: spack.platforms.test.Test()
+# FIXME: Since we change the architecture above, we have to (re)initialize
+# FIXME: the config singleton. If it gets initialized too early with the
+# FIXME: actual architecture, tests will fail.
+spack.config.config = spack.config._config()
 
 
 #

